### PR TITLE
Merge Analyzer changes into main branch

### PR DESCRIPTION
**Pull request by jchoi92k**
**Analyzer main update**
- Added custom LocalRecognizer for the third place approach.
- Restructured parts of code for efficiency and readability.
- Revised some of the rule-based approaches for more generalizability.
- Uses the configuration file in `configs` folder for configuration.
- Added comments and type hints for clarity
- TODO: there are still some one part of code where I think could be changed to make it more efficient

**Data source updates**
- Generated `.parquet` files for zip codes and first names and last names.
- Removed some names from the first and last name groups using the following logic:
- Only kept alphabets, spaces, and common accented characters (upper and lower `áàâäãåæçéèêëíìîïñóòôöõøùúûüýÿ`); this removed around 200k first/last names (kept around 730k).
- Config file points to some of these parquet files to help the custom recognizer load relevant files.

**Project structure updates**
- Changed GitIgnore to Ignore entire data folder
- Need to add data files and model files into respective folders (`data`, `models`) for code to run:
- I set up a folder on Google Drive to help us keep all files in the data folder in sync. Here's the link: https://drive.google.com/drive/folders/1Ru3bLULgt-FhqgaL90uoAV69pHOZ5el0?usp=sharing
- Also need to add the vectorizer models from the third place winning solution to the models folder: https://www.kaggle.com/code/devinanzelmo/piidd-efficiency-3rd-inference/input

**Pruning**
- Added static method `prune_results` to the custom analyzer class. Takes in list of `RecognizerResult`s and removes overlaps. 
- Can be changed later to use score-based approach. Currently just keeps the longer string.

**Misc**
- Slight changes to `app.py` to remove deprecation warnings

### CODE REVIEW

File: app.py [hunk 1]
```diff
@@ -17,9 +17,12 @@
 
 st.set_page_config(page_title="Student Name Detector (English)", layout="wide")
 
+# Pickled vectorizer freaks out if this isn't here
+def identity(x):
```
Comment (langdonholmes) [inline app.py:21]: lol. I can't imagine why this is necessary, but whatever...

**Review by langdonholmes — APPROVED** [review]
Hi Joon, thanks for your work here.

I haven't tested myself yet, but the changes look good.

Let's take a break on improving the analyzer for now. I think the next step should be to publish a pypi package with some basic usage instructions.

From there, we can work on building an executable distribution with a simple GUI. Let me know what library you think would be best. If you get the bones set up, I can probably work on fleshing things out when I have some more time (or funding).

If we have time, then we can work on either adding configuration options or improving the analyzer. I see two realistic options:
- Train our own deberta-xsmall using competition data + augmented data produced by Kagglers. We could then export this to ONXX and run it in the ONXX runtime (on CPU) in PIILO.
- Focus on the configuration options and keep the lightweight approach, but make names lists and what not configurable. In this way, we would kind of make the user responsible for getting the analyzer to work well with their data.

I do not think that we will make much progress creating a more generalizable pattern-based approach. These approaches are inherently brittle, and only work really well when they are catered to a specific dataset/context.

**Comment by jchoi92k** [comment]
I ended up doing what your proposed in the reverse order: I have a crude, alpha version executable packaged up. I used Pyinstaller. Will share more on Slack. I'll see if we can get this uploaded to Pypi.

File: app.py (hunk 1 above)
Comment (jchoi92k) [reply app.py:21]: Here's a reference:
https://rebeccabilbro.github.io/module-main-has-no-attribute/

Seems like a more fundamental fix would involve retraining or repackaging the model, and I didn't really want to spend time looking into this.

**Pull request merged**